File: Python/LCA.py
# Keira Gatt
# Student #: 19334557
# Date: 11.10.21
#
# Construct a Binary Search Tree & find lowest common ancestor for 2 given keys


class BSTNode:
    def __init__(self, val=None):  # Structure for node
        self.left = None  # Nodes to the left and right of itself
        self.right = None
        self.val = val  # Value in node
        self.size = 1  # Number of nodes in tis subtree (including itself)

    # Return number of key-value pairs in BST
    def bst_size(self):
        if self.val is None:
            return 0
        else:
            return self.size

    # Check if tree is empty (size = 0)
    def is_empty(self):
        return self.bst_size() == 0

    # Insert new node into BST
    def insert(self, val):
        if not self.val:  # If root is null -> set new value as root
            self.val = val
            self.size = 1
            return

        if self.val == val:  # If the value already exists in BST don't add new node
            return

        if val < self.val:  # If new value less than current node -> go left
            if self.left:  # If tree has a left subtree go left
                self.left.insert(val)
            else:  # Else create a left subtree
                self.left = BSTNode(val)

        if val > self.val:  # If new value less than current node -> go right
            if self.right:  # If tree has a right subtree go left
                self.right.insert(val)
            else:  # Else create a right subtree
                self.right = BSTNode(val)

        # Set size of tree
        if self.left is not None and self.right is not None:  # If left & right subtree exist
            self.size = 1 + self.left.bst_size() + self.right.bst_size()
        elif self.right is not None:  # If only right subtree exist
            self.size = 1 + self.right.bst_size()
        else:  # If only left subtree exist
            self.size = 1 + self.left.bst_size()

        return

    # Call recursive method to print BST inorder & format output string
    def inorder(self):
        if self.is_empty():  # Check for empty tree
            print("Can't print an empty tree!")
            return -1

        rec_res = self.inorder_rec(self.val, []).copy()  # Call Recursive inorder method & put result in array

        # Format output String
        output_res = [None] * ((len(rec_res) * 2) - 1)
        flag = 0
        a = 0

        for x in range(0, (len(rec_res) * 2) - 1):
            if flag == 0:
                output_res[x] = rec_res[a]
                a += 1
                flag = 1
            else:
                output_res[x] = ", "
                flag = 0

        return ''.join(output_res)

    def inorder_rec(self, x, str_buffer):

        if self is not None:

            if self.left is not None:
                self.left.inorder_rec(x, str_buffer)

            str_buffer.append(str(self.val))

            if self.right is not None:
                self.right.inorder_rec(x, str_buffer)

        return str_buffer

    # Check if value exists in BST
    def exists(self, val):
        if val == self.val:  # If value == current node
            return True

        if val < self.val:  # If value < current node go left
            if self.left is None:  # If there is no left subtree return false
                return False
            return self.left.exists(val)  # Else go left

        # If value > current node go right
        if self.right is None:  # If there is no right subtree return false
            return False
        return self.right.exists(val)  # Else go right

    # Check if both values are in BST
    def val_check(self, val1, val2):
        if self.exists(val1):
            if self.exists(val2):
                return True
        return False

    # Find LCA for 2 given values
    def lca(self, n1, n2):
        flag = self.val_check(n1, n2)  # Check both values are in BST

        if not flag:
            print("Invalid number(s)")
            return -1

        # If both n1 and n2 are smaller than self, then LCA
        # lies in left
        if (self.val > n1 and self.val > n2):
            return self.left.lca(n1, n2)

        # If both n1 and n2 are greater than self, then LCA
        # lies in right
        if (self.val < n1 and self.val < n2):
            return self.right.lca(n1, n2)

        return self.val

File: Java/LCA.java
/*
 * Keira Gatt
 * Student #: 19334557
 * Date: 20.09.21
 * 
 */

/*
 * Construct a Binary Search Tree & find lowest common ancestor for 2 given keys
 */
public class LCA {

	
	class Node{ 														// Structure for node
		
        int key;														// Value in node
        Node left, right;												// Nodes to the left and right of itself
        int size;														// Number of nodes in its subtree
 
        public Node(int item){
        	
            key = item;
            left = right = null;
            
        }
    }
	
	Node root; 															// Root node
	
	LCA(){root = null;}
	
	public void insert(int key) { root = insert(root, key); } 			// Call recursive method to insert node in BST
	
	private Node insert(Node root, int key) {
		
		if(root == null) root = new Node(key);							// If root is null -> set new value as root
		else if(key < root.key) root.left = insert(root.left, key);		// If new value less than current node -> go left
		else if (key > root.key) root.right = insert(root.right, key);	// If new value less than current node -> go left
		
		root.size = 1 + size(root.left) + size(root.right);				// Get size of node by adding its left & right subtree to itself
		return root;
		
	}
	
	// Return number of key-value pairs in BST
    public int size() { return size(root); }
    
	// Return number of key-value pairs in BST rooted at x
    private int size(Node x) {
    	
        if (x == null) return 0;
        else return x.size;
        
    }
	
    public boolean isEmpty() { return size() == 0; }				// Check if tree is empty (size = 0)
    
	String inorder(){return inorderRec(root, "");}					// Call recursive method to print BST inorder
 
    String inorderRec(Node root, String strBuff) {					// Traverse BST inorder & print keys
    	
    	StringBuilder strBuilder = new StringBuilder(strBuff);
    	strBuilder = strBuilder.append("(");
    	
        if (root != null) {
        	
        	strBuilder.replace(0, strBuilder.length(), inorderRec(root.left, strBuilder.toString()));
        	strBuilder = strBuilder.append(root.key);	
        	strBuilder.replace(0, strBuilder.length(), inorderRec(root.right, strBuilder.toString()));
            
        }
        
        strBuilder = strBuilder.append(")");
        return strBuilder.toString();
        
    }
    
     int lca(int key1, int key2) {									// Call recursive method if both keys are valid
    																// Else return -1
    	if(keyChecker(key1,key2)) return (lca(root, key1, key2)).key;
    		
    	return -1;

    }
    
    private boolean keyChecker(int key1, int key2) { 					// Check if both keys are in BST
    	
    	if(isEmpty()) {
    		
    		System.out.println("Empty Tree!");
    		return false;
    	
    	}
    	
    	Node searchNode = search(key1);									// If first key not in tree return invalid
    	
    	if(searchNode == null) { 
    		
    		System.out.printf("Invalid key 1 : %d\n", key1);
    		return false;
    	
    	}
    	
    	searchNode = search(key2);										// If second key not in tree return invalid
    	
    	if(searchNode == null) { 
    		
    		System.out.printf("Invalid key 2 : %d\n", key2);
    		return false;
    	
    	}
    	else return true;												// Else search query is valid
    	
    	
    }

    private Node search(int key) {									// Search BST for key
    	
    	Node currentNode = root;
    	
    	while(currentNode.key != key) {
    		
    		if(currentNode != null) {

    			if(currentNode.key > key) currentNode = currentNode.left;
    			else currentNode = currentNode.right; 			// currentNode.key < key
    			
    		}
    		
    		if(currentNode == null) return null;
 
    	}
    	
    	return currentNode;
    	
    }
    
    private Node lca(Node node, int key1, int key2) {								// Starting from top of tree, traverse down to find LCA
    	
    	if(node.key > key1 && node.key > key2) return lca(node.left, key1, key2); 	// If current node is greater than both keys -> go left
    	
    	if(node.key < key1 && node.key < key2) return lca(node.right, key1, key2); 	// If current node is less than both keys -> go right
    	
    	return node;
    }
    
}

